src/OrchardCore.Modules/Orchard.Lucene/LuceneIndexManager.cs

```diff
@@ -0,0 +1,347 @@
+using Directory = System.IO.Directory;
+using LuceneNetCodecs = Lucene.Net.Codecs;
+
+namespace Orchard.Lucene
+{
+    /// <summary>
+    /// Provides methods to manage physical Lucene indices.
+    /// This class is provided as a singleton to that the index searcher can be reused across requests.
+    /// </summary>
+    public class LuceneIndexManager : IDisposable
+    {
+        private readonly IHostingEnvironment _hostingEnvironment;
+        private readonly string _rootPath;
+        private readonly DirectoryInfo _rootDirectory;
+
+        private ConcurrentDictionary<string, IndexReaderPool> _indexPools = new ConcurrentDictionary<string, IndexReaderPool>(StringComparer.OrdinalIgnoreCase);
+        private ConcurrentDictionary<string, IndexWriter> _writers = new ConcurrentDictionary<string, IndexWriter>(StringComparer.OrdinalIgnoreCase);
+
+        private static LuceneVersion LuceneVersion = LuceneVersion.LUCENE_48;
+
+        static LuceneIndexManager()
+        {
+            SPIClassIterator<LuceneNetCodecs.Codec>.Types.Add(typeof(LuceneNetCodecs.Lucene46.Lucene46Codec));
+            SPIClassIterator<LuceneNetCodecs.PostingsFormat>.Types.Add(typeof(LuceneNetCodecs.Lucene41.Lucene41PostingsFormat));
+            SPIClassIterator<LuceneNetCodecs.DocValuesFormat>.Types.Add(typeof(LuceneNetCodecs.Lucene45.Lucene45DocValuesFormat));
+        }
+
+        public LuceneIndexManager(
+            IHostingEnvironment hostingEnvironment,
+            IOptions<ShellOptions> shellOptions,
+            ShellSettings shellSettings
+            )
+        {
+            _hostingEnvironment = hostingEnvironment;
+
+            _rootPath = Path.Combine(
+                _hostingEnvironment.ContentRootPath,
+                shellOptions.Value.ShellsRootContainerName,
+                shellOptions.Value.ShellsContainerName,
+                shellSettings.Name, "Lucene");
+
+            _rootDirectory = Directory.CreateDirectory(_rootPath);
+        }
+
+        public void CreateIndex(string indexName)
+        {
+            var path = new DirectoryInfo(Path.Combine(_rootPath, indexName));
+
+            if (!path.Exists)
+            {
+                path.Create();
+            }
+
+            Write(indexName, _ => { }, true);
+        }
+
+        public void DeleteDocuments(string indexName, IEnumerable<string> contentItemIds)
+        {
+            Write(indexName, writer =>
+            {
+                writer.DeleteDocuments(contentItemIds.Select(x => new Term("ContentItemId", x)).ToArray());
+
+                writer.Commit();
+
+                if (_indexPools.TryRemove(indexName, out var pool))
+                {
+                    pool.MakeDirty();
+                }
+            });
+        }
+
+        public void DeleteIndex(string indexName)
+        {
+            lock (this)
+            {
+                if (_writers.TryRemove(indexName, out var writer))
+                {
+                    writer.Dispose();
```

Comment:
The more I think about it the more I think we can't fix it correctly. We need to expose the folder using a common web service for all tenants using the same filesystem. So either we do that and configure a tenant-node to use it, or we have a folder for each tenant-node. But in the latter solution we need to expose the Rebuild/Reset actions as a command that can be registered on all nodes. Or we can say that for multi-instance system we need to switch to a service based solution, like baked by SQL or Blob storage, or using Elastic search.

So for now I'd suggest we do nothing more. Like it's done in O1.